### PR TITLE
rt: reduce an unnecessary lock operation

### PR DESCRIPTION
**Pull request by biluohc**
## Motivation

I found an unnecessary lock operation. 
https://github.com/tokio-rs/tokio/blob/afd2189eec8b21075a98613ffc2e11c38b82d4b9/tokio/src/runtime/blocking/pool.rs#L222-L271

```rs
        let shutdown_tx = {
            let mut shared = self.inner.shared.lock(); 
           ...
       }

        if let Some(shutdown_tx) = shutdown_tx {
            let mut shared = self.inner.shared.lock();
       }
```

## Solution

delete it

### CODE REVIEW

File: tokio/src/runtime/blocking/pool.rs [hunk 1]
```diff
@@ -220,54 +220,51 @@ impl fmt::Debug for BlockingPool {
 
 impl Spawner {
     pub(crate) fn spawn(&self, task: Task, rt: &Handle) -> Result<(), ()> {
-        let shutdown_tx = {
-            let mut shared = self.inner.shared.lock();
+        let add_thread = |shared: &mut MutexGuard<'_, Shared>, shutdown_tx| {
+            if let Some(shutdown_tx) = shutdown_tx {
+                let id = shared.worker_thread_index;
```
Comment (Darksonn) [inline tokio/src/runtime/blocking/pool.rs:225]: You only call this once. Just put it where you use it.

Comment (biluohc) [reply tokio/src/runtime/blocking/pool.rs:225]: The original code use `let  shutdown_tx = {}` may be to consider that the let block needs to be called multiple times in the future?

And the static dispatch closure should have no additional overhead?

Comment (Darksonn) [reply tokio/src/runtime/blocking/pool.rs:225]: I'm not worried about overhead. I just think it would be easier to read the code without the closure.

Comment (biluohc) [reply tokio/src/runtime/blocking/pool.rs:225]: I don't think the closure makes code more difficult to read,  it's more clear.

Comment (biluohc) [reply tokio/src/runtime/blocking/pool.rs:225]: But I'll change it if you think it's better without  the closure..

**Comment by biluohc** [comment]
Please continue to review @Darksonn

**Comment by Darksonn** [comment]
It seems reasonable to me.

**Comment by biluohc** [comment]
ok, thanks you very much

**Review by carllerche — APPROVED** [review]
Thanks. I can't find a reason for the unlock/lock in the history or comments. Tests pass, so lets merge it.

**Pull request merged**